File: tests/drivers/javascript/v4_1/max_query_length.js
// Determines how long could be a query executed
// from JavaScript driver.
//
// Performs binary search until the maximum possible
// query size has found.

// init driver
var neo4j = require('neo4j-driver');
var driver = neo4j.driver("bolt://localhost:7687",
                          neo4j.auth.basic("", ""),
                          { encrypted: 'ENCRYPTION_OFF' });

// init state
var property_size = 0;
var min_len = 1;
var max_len = 1000000;

// hacking with JS and callbacks concept
function serial_execution() {
  var next_size = [Math.floor((min_len + max_len) / 2)];
  setInterval(function() {
    if (next_size.length > 0) {
      property_size = next_size.pop();
      var query = "CREATE (n {name:\"" +
        (new Array(property_size)).join("a")+ "\"})";
      var session = driver.session();
      session.run(query, {}).then(function (result) {
        console.log("Success with the query length " + query.length);
        if (min_len == max_len || property_size + 1 > max_len) {
          console.log("\nThe max length of a query from JS driver is: " +
            query.length + "\n");
          session.close();
          driver.close();
          process.exit(0);
        }
        min_len = property_size + 1;
        next_size.push(Math.floor((min_len + max_len) / 2));
      }).catch(function (error) {
        console.log("Failure with the query length " + query.length);
        max_len = property_size - 1;
        next_size.push(Math.floor((min_len + max_len) / 2));
      }).then(function(){
        session.close();
      });
    }
  }, 100);
}

// execution
console.log("\nDetermine how long can be a query sent from JavaScript driver.");
serial_execution(); // I don't like JavaScript

File: tests/drivers/python/v4_1/max_query_length.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Copyright 2021 Memgraph Ltd.
#
# Use of this software is governed by the Business Source License
# included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
# License, and you may not use this file except in compliance with the Business Source License.
#
# As of the Change Date specified in that file, in accordance with
# the Business Source License, use of this software will be governed
# by the Apache License, Version 2.0, included in the file
# licenses/APL.txt.

from neo4j import GraphDatabase, basic_auth

driver = GraphDatabase.driver("bolt://localhost:7687",
                              auth=basic_auth("", ""),
                              encrypted=False)

query_template = 'CREATE (n {name:"%s"})'
template_size = len(query_template) - 2  # because of %s
min_len = 1
max_len = 1000000

# binary search because we have to find the maximum size (in number of chars)
# of a query that can be executed via driver
while True:
    assert min_len > 0 and max_len > 0, \
        "The lengths have to be positive values! If this happens something" \
        " is terrible wrong with min & max lengths OR the database" \
        " isn't available."
    property_size = (max_len + min_len) // 2
    try:
        driver.session().run(query_template % ("a" * property_size)).consume()
        if min_len == max_len or property_size + 1 > max_len:
            break
        min_len = property_size + 1
    except Exception as e:
        print("Query size %s is too big!" % (template_size + property_size))
        max_len = property_size - 1

assert property_size == max_len, "max_len probably has to be increased!"

print("\nThe max length of a query from Python driver is: %s\n" %
      (template_size + property_size))

# sessions are not closed because all sessions that are
# executed with wrong query size might be broken
driver.close()
